### PR TITLE
Add %f to datetime.strftime()

### LINKED ISSUE #16834: DateTime.strftime ignoring `%f` modifier
As [reported on Discourse](https://chapel.discourse.group/t/f-format-does-not-seem-to-work-with-datetime/2212), our `strftime()` routine does not recognize `"%f"` even though our pointer to the Python documentation implies that it probably should.  For example:

```chapel
use DateTime;
var now = datetime.now();
writeln(now.strftime("%H:%M:%S.%f"));
writef("%06i\n",now.microsecond);
```

prints:

```
18:47:11.%f
```

### PR DESCRIPTION
**Pull request by leekillough**
Fixes #16834

### CODE REVIEW

File: modules/standard/DateTime.chpl [hunk 1]
```diff
@@ -1197,7 +1197,36 @@ module DateTime {
     timeStruct.tm_wday = (weekday(): int(32) + 1) % 7; // shift Sunday to 0
     timeStruct.tm_yday = (this.replace(tzinfo=nil) - new datetime(year, 1, 1)).days: int(32);
 
-    strftime(c_ptrTo(buf), bufLen, fmt.c_str(), timeStruct);
+    // Iterate over format specifiers in strftime()
+    iter strftok(s)
```
Comment (daviditen) [inline modules/standard/DateTime.chpl:1201]: Could this use `string.replace()` instead of walking the characters?

```chapel
const micro = "%06i".format(chpl_time.chpl_microsecond);
const refmt = s.replace("%f", micro);
```

File: modules/standard/DateTime.chpl [hunk 2]
```diff
@@ -1197,7 +1197,36 @@ module DateTime {
...
+      for c in s {
+        if per == "" {
+          if c == '%' {
+             per += c;
+          } else {
+             yield c;
+          }
+        } else {
+          per += c;
+          if c != '-' {
```
Comment (daviditen) [inline modules/standard/DateTime.chpl:1213]: What format is this line looking for/trying to build? `"%-?"`?

File: modules/standard/DateTime.chpl (hunk 1 above)
Comment (leekillough) [reply modules/standard/DateTime.chpl]: > Could this use `string.replace()` instead of walking the characters?
> 
> ```chapel
> const micro = "%06i".format(chpl_time.chpl_microsecond);
> const refmt = s.replace("%f", micro);
> ```

No, because `%%f` needs to be printed as `%f` instead of as `%012345`.

I added a test case with `%%f` to test this case:
```chapel
assert(t.strftime("%m %d %y %S %M %H %%f %j") == "12 31 04 33 22 06 %f 366");
```

File: modules/standard/DateTime.chpl (hunk 2 above)
Comment (leekillough) [reply modules/standard/DateTime.chpl:1213]: > What format is this line looking for/trying to build? `"%-?"`?

`strftime()` allows all numerical output specifiers to use an optional `-` modifier, which specifies that instead of 0-padding the number so that it has a constant number of digits, the number will be formatted in a platform-specific way, which in GNU libc, does not include any leading spaces or 0s as padding. So we parse for `-` after `%` as an optional modifier. We do not need to handle arbitrary widths or precisions like in `printf()`, however.

Comment (leekillough) [reply modules/standard/DateTime.chpl]: I extended it to allow `%-*` (no padding), `%_*` (space padding), `%0*` (explicit 0 padding), and `%E*` and `%O*` (POSIX extensions), which are all allowed as modifiers by the C `strftime()`, since we are calling C `strftime()` directly.

See `man strftime` for the explanation of these modifiers.

File: .dir-locals.el [hunk 3]
```diff
@@ -0,0 +1,23 @@
+;; This causes emacs to abide by formatting rules: No tab characters, except
```
Comment (daviditen) [inline .dir-locals.el:1]: This seems unrelated and should maybe be in a different PR.

**Review by daviditen — APPROVED** [review]
I'm not crazy about adding the emacs change in this PR, but the rest looks good to me.

**Comment by leekillough** [comment]
> I'm not crazy about adding the emacs change in this PR, but the rest looks good to me.

I was getting CI errors because of tabs. I can submit another PR for the `.dir-locals.el` addition, which untabifies and gets rid of common newline and trailing space errors at file save time.

File: .gitignore [hunk 4]
```diff
@@ -99,6 +99,8 @@ tags
 /test/**/compperfdat
 /test/**/perfdat
 
+testDatetime
```
Comment (bradcray) [inline .gitignore:102]: @leekillough / @daviditen : It doesn't seem to me that this change should be necessary (i.e., we don't list all test names in `.gitignore` (that goodness)).  Was it?

**Comment by bradcray** [comment]
@leekillough : Congratulations on merging your first PR!

As a few points of order, when merging PRs, we:

* include more narrative in the PR about what was done and how (see other recently merged PRs for examples)

* cut and paste this same text into the merge message using the format:

```
One-line headline for the PR

[reviewed by @daviditen]

<cut-and-paste of PR text>
```

* when merging, don't use the squash-and-merge feature in Github because it (a) removes the DCO sign-off lines and (b) bypasses our mechanisms that generate Discourse topics for each commit.

Details on all of this are here: https://github.com/chapel-lang/chapel/blob/master/doc/rst/developer/bestPractices/ContributorInfo.rst#pull-request-description

File: .gitignore (hunk 4 above)
Comment (mppf) [reply .gitignore:102]: I also don't think this line should be added to `.gitignore`.
